core/src/services/dbfs/core.rs

```diff
@@ -0,0 +1,237 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::fmt::Debug;
+
+use base64::prelude::BASE64_STANDARD;
+use base64::Engine;
+use bytes::Bytes;
+use http::header;
+use http::Request;
+use http::Response;
+use http::StatusCode;
+use serde_json::json;
+
+use crate::raw::*;
+use crate::*;
+
+use super::error::parse_error;
+
+pub struct DbfsCore {
+    pub root: String,
+    pub endpoint: String,
+    pub token: String,
+    pub client: HttpClient,
+}
+
+impl Debug for DbfsCore {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("DbfsCore")
+            .field("root", &self.root)
+            .field("endpoint", &self.endpoint)
+            .field("token", &self.token)
+            .finish_non_exhaustive()
+    }
+}
+
+impl DbfsCore {
+    pub async fn dbfs_create_dir(&self, path: &str) -> Result<Response<IncomingAsyncBody>> {
+        let url = format!("{}/api/2.0/dbfs/mkdirs", self.endpoint);
+        let mut req = Request::post(&url);
+
+        let auth_header_content = format!("Bearer {}", self.token);
+        req = req.header(header::AUTHORIZATION, auth_header_content);
+
+        let p = build_rooted_abs_path(&self.root, path)
+            .trim_end_matches('/')
+            .to_string();
+
+        let req_body = format!("{{\"path\": \"{}\"}}", percent_encode_path(&p));
```

Comment:
Please don't build json by hand. Use `serde_json` instead.

Reply:
Done. I don't know why my original thought was that if there's only one element, then using string formatting would be cleaner.